### PR TITLE
core: warn if a generator is world-writable

### PR DESCRIPTION
**Pull request by lnykryn**
... because that is obviously a security risk.

### CODE REVIEW

File: src/shared/exec-util.c [hunk 1]
```diff
@@ -156,6 +156,17 @@ static int do_execute(
...
 
+                if (FLAGS_SET(flags, EXEC_DIR_WARN_WORLD_WRITABLE)) {
+                        struct stat st;
+
+                        r = stat(t, &st);
+
+                        if (r < 0)
+                                log_warning_errno(errno, "Failed to stat %s: %m", t);
+                        else if (S_ISREG(st.st_mode) && (st.st_mode & 0002))
+                                log_warning("%s is world writable. This is not a good idea.", t);
```
Comment (bluca) [inline src/shared/exec-util.c]: Maybe expand a little to match a similar warning that already exists elsewhere: `"%s is marked world-writable, which is a security risk as it is executed with privileges. Please remove world writability permission bits. Proceeding anyway.`

File: src/shared/exec-util.c [hunk 2]
```diff
@@ -156,6 +156,17 @@ static int do_execute(
                         log_debug("About to execute %s%s%s", t, argv ? " " : "", argv ? strnull(args) : "");
                 }
 
+                if (FLAGS_SET(flags, EXEC_DIR_WARN_WORLD_WRITABLE)) {
+                        struct stat st;
+
+                        r = stat(t, &st);
+
```
Comment (bluca) [inline src/shared/exec-util.c]: extra empty line

**Review by bluca — APPROVED** [review]

**Review by YHNdnzj — COMMENTED** [review]
Hmm, I don't really see the point of this? Every system dir/config file writable by others is a security risk, so what's special about generators? IOW, this seems too basic that I doubt such warning would provide help?

**Comment by bluca** [comment]
We have some warnings like this for selected items. Generators are dropped in by users, so it is not inconceivable to make such a mistake, and it is really damaging and hard to notice otherwise, so it seems worth having

**Comment by lnykryn** [comment]
> Hmm, I don't really see the point of this? Every system dir/config file writable by others is a security risk, so what's special about generators? IOW, this seems too basic that I doubt such warning would provide help?

I think it will. We found out that our users often even overreact to warning messages in the logs. Also this was a recommendation from our security team after an internal audit of systemd.

**Comment by YHNdnzj** [comment]
> We have some warnings like this for selected items. Generators are dropped in by users, so it is not inconceivable to make such a mistake, and it is really damaging and hard to notice otherwise, so it seems worth having

My argument is that the same applies to anything, i.e. the whole /etc/. Unit files with write access can be at least equally dangerous.

**Comment by lnykryn** [comment]
https://github.com/systemd/systemd/blob/main/src/basic/fs-util.c#L346

**Comment by bluca** [comment]
> > We have some warnings like this for selected items. Generators are dropped in by users, so it is not inconceivable to make such a mistake, and it is really damaging and hard to notice otherwise, so it seems worth having
> 
> My argument is that the same applies to anything, i.e. the whole /etc/. Unit files with write access can be at least equally dangerous.

Unit files are a bit more obvious as they are config files so you see what they do. Executables are much more opaque

**Review by yuwata — APPROVED** [review]
LGTM.

**Pull request merged**